### PR TITLE
Handle monitor enter/exit on value based instances in Power

### PR DESCRIPTION
**Pull request by a7ehuo**
Depends on 
- [x] #11358

Related to JEP390

Signed-off-by: Annabelle Huo <Annabelle.Huo@ibm.com>

### CODE REVIEW

**Comment by a7ehuo** [comment]
@gita-omr Ready to review, thanks!
@hzongaro FYI

**Comment by gita-omr** [comment]
I am a bit puzzled since I think we already did this work. But will need to look after Jan 4th.

**Comment by a7ehuo** [comment]
The work done in #10214 is to check if monitor enter/exit is on `value type` instances. This change is to check if monitor enter/exit is on `value based` class objects.  A new `@ValueBased` annotation will be introduced in Java 16 as described in #10620.  This change combines the check on `value type` instances and `value based` class objects.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 1]
```diff
@@ -7727,11 +7729,13 @@ void J9::Power::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(TR::N
...
    TR::MemoryReference *classFlagsMemRef = TR::MemoryReference::createWithDisplacement(cg, objectClassReg, static_cast<uintptr_t>(fej9->getOffsetOfClassFlags()), 4);
 
-   //check J9ClassIsValueType flag
-   generateTrg1MemInstruction(cg,TR::InstOpCode::lwz, node, tempReg, classFlagsMemRef);
-   generateTrg1Src1ImmInstruction(cg, TR::InstOpCode::andi_r, node, tempReg, tempReg, condReg, J9ClassIsValueType);
+   //check J9_CLASS_DISALLOWS_LOCKING_FLAGS (J9ClassIsValueType | J9ClassIsValueBased)
+   generateTrg1MemInstruction(cg, TR::InstOpCode::lwz, node, temp1Reg, classFlagsMemRef);
 
-   //If obj is value type, call VM helper and throw IllegalMonitorState exception, else continue as usual
+   loadConstant(cg, node, classFlag, temp2Reg);
```
Comment (gita-omr) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7735]: The fact that the new flag is outside of the immediate value range for Power is a bit concerning but I guess we ran out of smaller flags? @tajila

**Comment by gita-omr** [comment]
@IBMJimmyk could you please review, especially on the subject of the temporary register being reused?

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp (hunk 1 above)
Comment (tajila) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7735]: Yes, the smaller flags are all used up. How large is the immediate range on power?

Comment (gita-omr) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7735]: @tajila it’s 16 bits.

File: runtime/compiler/p/codegen/J9TreeEvaluator.cpp [hunk 2]
```diff
@@ -7718,7 +7720,7 @@ static bool simpleReadMonitor(TR::Node *node, TR::CodeGenerator *cg, TR::Node *o
    return true;
    }
 
-void J9::Power::TreeEvaluator::generateCheckForValueTypeMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol *helperCallLabel, TR::Register *objReg, TR::Register *objectClassReg, TR::Register *tempReg, TR::Register *condReg, TR::CodeGenerator *cg)
+void J9::Power::TreeEvaluator::generateCheckForValueMonitorEnterOrExit(TR::Node *node, TR::LabelSymbol *helperCallLabel, TR::Register *objReg, TR::Register *objectClassReg, TR::Register *temp1Reg, TR::Register *temp2Reg, TR::Register *condReg, TR::CodeGenerator *cg, int32_t classFlag)
```
Comment (IBMJimmyk) [inline runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7723]: Was there a reason for adding the `int32_t classFlag` parameter? It looks like the constant `J9_CLASS_DISALLOWS_LOCKING_FLAGS` is the only value that is ever passed in.

Comment (a7ehuo) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7723]: The purpose of adding a `classFlag` parameter is to make `generateCheckForValueTypeMonitorEnterOrExit()` more generic for future flag changes, although at the moment only  `J9_CLASS_DISALLOWS_LOCKING_FLAGS` is used.

Comment (IBMJimmyk) [reply runtime/compiler/p/codegen/J9TreeEvaluator.cpp:7723]: Okay, that makes sense.

**Comment by IBMJimmyk** [comment]
The only comment I had was addressed so this looks good to me.

**Review by IBMJimmyk — APPROVED** [review]

**Comment by gita-omr** [comment]
Jenkins test sanity aix,plinux jdk8,jdk11

**Comment by a7ehuo** [comment]
@gita-omr All tests passed

**Pull request merged**